enhancements/olm/olm-managed-operator-metrics.md

```diff
@@ -0,0 +1,145 @@
+
+This feature present unique value on OpenShift Clusters, where Telemeter could provide Red Hat with the ability to identify and proactively fix customer issues caused by Red Hat owned operators.
+
+### Goals
+
+- Provide operators with a first-class way of creating `ServiceMonitors` objects
+- Provide operators with a first-class way of creating `PrometheusRules` objects
+- Provide operators with a first-class way of specifying a namespace that they must be deployed to
+
+### Non-Goals
+
+- Provide generic metrics about OLM managed operators
+
+## Proposal
+
+### OLM support for new object in bundles
+
+The bundle object currently has the following format:
+
+```bash
+$ tree
+bundle
+    ├── manifests
+    │   ├── example.crd.yaml
+    │   ├── example.csv.yaml
+    │   └── example.rbac.yaml
+    └── metadata
+        └── annotations.yaml
+```
+
+Notice that OLM understands how to create and manage CSVs, CRDs, and RBAC files. This enhancement proposes expanding this feature to support `Namespaces`, `ServiceMonitors` and `PrometheusRules` objects:
+
+```bash
+$ tree
+bundle
+    ├── manifests
+    │   ├── example.crd.yaml
+    │   ├── example.csv.yaml
+    │   ├── example.ns.yaml
+    │   ├── example.prometheusrule.yaml
+    │   ├── example.rbac.yaml
+    │   └── example.servicemonitor.yaml
+    └── metadata
+        └── annotations.yaml
+```
+
+With this change, OLM will need to be updated to manage the `Namespace`, `ServiceMonitor` and `PrometheusRule` objects through the lifecycle of an operator.
+
+#### Namespace Support
+
+As OLM gains support for `Namespace` objects, The `ClusterServiceVersion` will be updated to include an optional `NamespaceRequirements` object field. The `NamespaceRequirements` will look like the following:
+
+```golang
+type NamespaceRequirements struct {
+  //+optional
+  // name represents where the operator must be deployed.
+  name string
+
+  //+optional
+  // labels are labels that must be present on the given namespace.
+  labels []string
+}
+```
+
+>Note: If `NamespaceRequirements.namespace` is set, the UI will only show the specified namespace as an install option.
+
+#### ServiceMonitor and PrometheusRule Support
+
+As OLM gains support for these Prometheus objects, there is a chance that installed operators may overload the Prometheus instance. It is important that a Cluster Admin is alerted that an operator will report metrics prior to deploying the operator. As such, we propose adding a `monitoringSupported` boolean field to the `ClusterServiceVersion` object. OLM will not install either of the Prometheus object unless the `monitoringSupported` field is set to true. Additionally, Cluster Admins will need to manually the operator's `InstallPlans` in the following scenarios:
```

Comment:
_Note: Continuing the conversation from https://github.com/operator-framework/enhancements/pull/4_
From @ecordell:

> This seems like an alternative the first thing, right? If I include a namespace name + the right labels, then it doesn't matter if "monitoring is supported" or not?

Reply:
@ecordell the namespace and labels could be used for purposes other than metrics - we don't have a way to identify this on upstream.

Reply:
Can someone explain to me why we should not install these resources based on the operator author's decision? If they don't want to include metrics then they should not. If Prometheus is not installed in the cluster, then OLM *should* skip those resources until they are available.

Reply:
@ecordell the suggested approach would work for integrating operators with the monitoring Prometheus operator but I do not know how this could apply upstream clusters.

If we want Cluster Admins to be notified that they are installing an operator that will share metrics with prometheus instances, and require them to approve the `InstallPlan`, I believe the workflow should be the same for both OpenShift and Upstream clusters.